### ISSUE TITLE
Is it possible to change grant_type to "authorization_code" while using signinSilent() ?

### ISSUE THREAD

**Issue by wingosite** [authts/oidc-client-ts#1239]
Any way to change the grant_type?

Please help me out. Thanks in advance...!!

**Comment by pamapa** [comment]
`signinSilent` supports only renew via iframe or refresh_token,
See https://github.com/authts/oidc-client-ts/blob/909df5752a7c740a8904bd7760f7683ce2c90bfd/src/UserManager.ts#L254-L298

**Comment by lmarquinez** [comment]
I'm trying to implement the silent refresh with the `authorization code` using the `oidc-client-ts` library in Angular 15, but I can't use the `offline_access` to get the `refresh_token` back as it keeps asking me to accept the consent each time a 
log in the aplication ([https://docs.duendesoftware.com/identityserver/v6/tokens/refresh/#consent](https://docs.duendesoftware.com/identityserver/v6/tokens/refresh/#consent)).
[https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess](https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess)

So I am trying to do it via an iframe as you mention. I have the following silent-refresh.html:

```
<head>
  <title></title>
</head>
<body>
  <script src="oidc-client-ts.js"></script>
  <script>
    new Oidc.UserManager().signinSilentCallback().catch((err) => {
      console.log(err);
    });
  </script>
</body>
```

And in my `auth.service.ts` I use the `signinSilent` and I also give this configuration:

```
export class AuthService {
  private manager: UserManager;
  private user: User = null;
  web = new WebStorageStateStore({});

  constructor(
    private router: Router,
  ) {
    this.manager = new UserManager({
      authority: authConfig.authorityUri,
      client_id: authConfig.clientId,
      redirect_uri: authConfig.redirectUri,
      post_logout_redirect_uri: authConfig.postLogoutRedirectUri,
      response_type: 'code',
      scope: 'openid profile',
      filterProtocolClaims: true,
      loadUserInfo: true,
      automaticSilentRenew: false,
      includeIdTokenInSilentRenew: true,
      silent_redirect_uri: 'http://localhost:4200/silent-refresh.html',
    } as UserManagerSettings);

    this.manager.getUser().then((user) => {
      this.user = user;
    });

    if (authConfig.silent_renew) {
      this.manager.events.addAccessTokenExpired(() => this.refreshToken());
    }
  }

  refreshToken(): void {
    if (!this.user) {
      return;
    }

    this.manager
      .signinSilent()
      .then((user) => {
        this.user = user;
      })
      .catch(() => {
        this.logout();
      });
  }
```

So when the access_token is expired, it makes the call to `authorize` with the following parameters:

```
client_id: xxx
redirect_uri: http://localhost:4200/silent-refresh.html
response_trype: code
scope: openid profile
state: xxx
code_callenge: xxx
code_challenge_method: S256
response_mode: query
prompt: none
id_token_hint: xxx
```

and then also makes the call to silent refresh with this parameters:

```
code: xxx
scope: openid profile
state: xxx
sesion_state: xxx
iss: xxx
```

From there it should make the post request where it returns the access_token, but it doesn't do it, so the session ends. 

When I was using the `oidc-client` library with the `implicit flow`, it did the following:
1. Ask for token with `/connect/authorize` and the callback leads to silent-refresh.html.
2. This file contains the library and already changes the old token for the other one.

I understand that what it has to do now with the `authorization code` is:
1. Ask for the code (which it does) with the request `/connect/authorize`, and the callback leads to silent-refresh.html.
2. Perform the request to get token with the above code. Which in this almost silent-refresh does not. The request should go to `/connect/token`.
3. Validate the token and start using it, removing the previous one. This also has to be done by the library.

**Comment by pamapa** [comment]
Some points:
- Instead of `signinSilentCallback` you can use `signinCallback`. This handles redirect, popup and silent callbacks
- `automaticSilentRenew: false,` ->`true`, this setting would enable handling the renew within the library
- `implicit flow` is not used by this library, this library used `code flow` with PKCE

**Comment by lmarquinez** [comment]
Yes, I am not using the implicit flow now, I was saying that with the previous package, `oidc-client` which was using the `implicit flow`, the silent-refresh through the iframe was working correctly. And after migrating to `oidc-client-ts` to `authorice code flow` it does not work.

I tried to change the `signinSilentCallback() ` from `silent-refresh.html` to `signinCallback()` and in the configuration I changed the `automaticSilentRenew ` from `false ` to `true ` and it doesn't work.

I also have this error in the browser console: `Uncaught ReferenceError: Oidc is not defined at VM216 silent-refresh.html`

**Comment by pamapa** [comment]
> I also have this error in the browser console: `Uncaught ReferenceError: Oidc is not defined at VM216 silent-refresh.html`

That is your issue, it seems to not able to be able to access the object `Oidc`. This is you instance of the `UserManager`, somewhere in your typescript/javascript code you create that manager, you must make sure to have access to it from the html file...

**Comment by lmarquinez** [comment]
In the end I got it to work but I noticed that every time it refreshes, I mean, every time you hit F5 for example, it asks for a new token again. Is it possible to avoid this request?
